README.md
# REST API for a school database

In this project for the Treehouse techdegree, I created a REST API using Express. The API provides a way for users to administer a school database containing information about courses: users can interact with the database by retrieving a list of courses, as well as adding, updating and deleting courses in the database.

In addition, users are required to create an account and log-in to make changes to the database.

## Technologies Used

- Node.js
- Express
- Sequelize ORM
- SQL
- REST API design
- JavaScript

## Installation

- Download or clone from Github
- Run `npm install`
- Run `npm run seed`
- Run `npm run start` to run the app

## Available npm scripts

The environment variable `RESTAPI_DB` must be set to the SQLITE database `./fsjstd-restapi.db`. This is already configured in `nodemon.json` for nodemon setups.  

### `npm run seed`

- Seeds the SQLITE database

### `npm start`

- Runs app via nodemon

### `npm run debug`

- Runs app via nodemon in inspect mode for debugging in Visual Studio Code or Google Chrome

## REST API Endpoints

### User

- `GET /api/users 200` - Returns the currently authenticated user
- `POST /api/users 201` - Creates a user, sets the Location header to "/", and returns no content

### Course

- `GET /api/courses 200` - Returns a list of courses (including the user that owns each course)
- `GET /api/courses/:id 200` - Returns the course (including the user that owns the course) for the provided course ID
- `POST /api/courses 201` - Creates a course, sets the Location header to the URI for the course, and returns no content (Authentication required)
- `PUT /api/courses/:id 204` - Updates a course and returns no content (Authentication required)
- `DELETE /api/courses/:id 204` - Deletes a course and returns no content  (Authentication required)
- The `PUT /api/courses/:id` and `DELETE /api/courses/:id` routes return a 403 status code if the current user doesn't own the requested course

### General remarks

- The following properties are filtered out: `password`, `createdAt`, `updatedAt`
- If a course or a route isn't found a 404 status is returned and no content

### Tests

- The [RESTAPI.postman_collection.json](./tests/RESTAPI.postman_collection.json) file is a collection of Postman requests that can be used to test the REST API.
- The [api.http](./tests/api.http) file serves the same purpose and can be used with the Visual Studio Code [REST Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) extension. It uses variables, so it can also be used to test a deployed version.
- The [gen-data.js tool](./helpers/README.md) in `./helpers` can be used to generate test data.

## Error Messages

- Error messages are always returned in an array of error objects

### Validation

- User: `firstName`, `lastName`, `emailAddress`, `password` are required
- Course: `title`, `description`, `userId` are required
- `emailAddress` must be unique and a valid email address
- `password` must be at least 8 characters long

### Validation Errors

- Validation errors are routed through from Sequelize and return the properties: `message`, `type`, `path`, `value`. This enables a client application to create their own error messages if required.

Example:

```json
{
  "errors": [
    {
      "message": "User.password cannot be null",
      "type": "notNull Violation",
      "path": "password",
      "value": null
    }
  ]
}
```

### Other Errors

- Other errors return the properties: `message`, `type`

Example:

```json
{
  "errors": [
    {
      "message": "Unexpected token f in JSON at position 5",
      "type": "SyntaxError"
    }
  ]
}
```


CHANGELOG.md

1.1.0 / 2020-02-07
==================

* update npm scripts
* add hosted api loc to api.http
* prepare for heroku deploy
* unify return json for getCourse(s) route
* remove temporary test data
* update validation messages
* update seed script
* add cors support
* add own package.json to ./helpers
* add eslint, prettier config


lib/utils.js
const auth = require('basic-auth');
const bcrypt = require('bcryptjs');
const db = require('../db');

const { Sequelize, attrib } = db;
const { User, Course } = db.models;

/**
 * Filters out properties in an object
 * @param {object} obj - object to filter
 * @param  {...string} props - properties to exclude
 * @returns {object} filtered object
 */
const filterObject = (obj, ...props) => {
  return Object.keys(obj).reduce((newObj, key) => {
    if (!props.includes(key)) {
      newObj[key] = obj[key];
    }
    return newObj;
  }, {});
};

/**
 * Filters Sequelize error properties
 * @param {object} err - Sequelize error object
 * @returns {Array} Array with errors to be returned as json
 */
const validationError = (err, req) => {
  // fields to be returned as json
  const ERR_FIELDS = ['message', 'type', 'path', 'value'];
  let errors;

  // Can only happen when adding/updating a course
  if (err instanceof Sequelize.ForeignKeyConstraintError) {
    errors = {
      message: 'user with "userId" doesn\'t exist as a user',
      type: 'Foreign Key Constraint Error',
      path: 'userId',
      value: req.body.userId,
    };
  }

  // handle all other validation errors
  if (err.errors) {
    errors = err.errors.map((error) => {
      return Object.keys(error).reduce((newObj, key) => {
        if (ERR_FIELDS.includes(key)) {
          newObj[key] = error[key];
        }
        // filter out password value if necessary
        if (newObj.path === 'password') {
          return filterObject(newObj, 'value');
        }
        return newObj;
      }, {});
    });
  }
  return { errors };
};

/**
 * Handler function to wrap each route with error handling
 * @param {function} cb - callback
 * @returns {function} function
 */
const asyncHandler = (cb) => {
  // eslint-disable-next-line consistent-return
  return async (req, res, next) => {
    try {
      await cb(req, res, next);
    } catch (error) {
      // Checks if error was created by Sequelize
      // https://github.com/sequelize/sequelize/blob/3e5b8772ef75169685fc96024366bca9958fee63/lib/errors.js
      if (error instanceof Sequelize.BaseError) {
        res.status(400).json(validationError(error, req));
      } else {
        return next(error);
      }
    }
  };
};

/**
 * Handles courses PUT/DELETE routes
 * @param {object} req - Express request object
 * @param {object} res - Express response object
 * @param {function} next - next middleware function
 * @param {function} cb - callback cb(course: Course), must return true on success
 */
// eslint-disable-next-line consistent-return
const courseHandler = async (req, res, next, cb) => {
  try {
    const { id } = req.params;
    const { emailAddress } = req.currentUser;

    const course = await Course.findByPk(id);
    if (course) {
      // only allow authenticated user to update/delete their own course(s)
      const user = await User.findByPk(course.userId);
      if (user.emailAddress === emailAddress) {
        if (await cb(course)) {
          res.status(204).end();
        }
      } else {
        res.status(403).end();
      }
    } else {
      res.status(404).end();
    }
  } catch (error) {
    if (error instanceof Sequelize.BaseError) {
      res.status(400).json(validationError(error, req));
    } else {
      return next(error);
    }
  }
};

/**
 * Middleware function to authenticate user via basic auth
 * @param {object} req - Express request object
 * @param {object} res - Express response object
 * @param {function} next - next middleware function
 */
const authenticateUser = async (req, res, next) => {
  let message = null;

  // Get the user's credentials from the Authorization header.
  const credentials = auth(req);

  if (credentials) {
    // Look for a user whose email matches the credentials `name` property.
    credentials.name = credentials.name.toLowerCase();
    const user = await User.findOne({
      attributes: [...attrib.user, 'password'],
      where: { emailAddress: credentials.name },
    });

    if (user) {
      const authenticated = bcrypt.compareSync(credentials.pass, user.password);
      if (authenticated) {
        console.log(
          `Authentication successful for username: ${user.emailAddress}`
        );

        // Store the user on the Request object.
        req.currentUser = filterObject(user.get({ plain: 'true' }), 'password');
      } else {
        message = `Authentication failure for username: ${user.emailAddress}`;
      }
    } else {
      message = `User not found for email address: ${credentials.name}`;
    }
  } else {
    message = 'Auth header not found';
  }

  if (message) {
    console.warn(message);
    res.status(401).end();
  } else {
    next();
  }
};

module.exports = {
  authenticateUser,
  filterObject,
  asyncHandler,
  validationError,
  courseHandler,
};
